# Challenge 3

## 1. Introduccion

Para superar la limitación de hierro, los microorganismos han desarrollado diversos mecanismos de adquisición, uno de los cuales es la producción de sideróforos. Los sideróforos son moléculas pequeñas de naturaleza quelante que tienen la capacidad de unirse al hierro y transportarlo al interior de la célula.

### Problematica

El hierro es un elemento esencial para la vida, pero su disponibilidad en la naturaleza es limitada. Los microorganismos han desarrollado mecanismos para adquirir hierro del entorno, uno de los cuales es la producción de sideróforos. Los sideróforos son moléculas pequeñas que quelan el hierro y lo transportan al interior de la célula.

La base de datos de sideróforos proporciona información valiosa sobre estos compuestos, incluyendo su estructura, propiedades y microorganismos asociados. Esta información puede ser utilizada para:

- Comprender mejor los mecanismos de adquisición de hierro en microorganismos.
- Desarrollar nuevas estrategias para combatir las enfermedades infecciosas.
- Diseñar sideróforos artificiales para aplicaciones biotecnológicas.


## 2. Metodologia

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
# Importamos los datos
df_db = pd.read_csv("https://raw.githubusercontent.com/inefable12/siderophores_database/main/Siderophore_DB.csv")
df_db1 = pd.read_csv("https://raw.githubusercontent.com/inefable12/siderophores_database/main/Siderophore_DB1.csv")

# Selecconamos solo los grupos que queremos
df_db_grupos = df_db[['hydroxamate','catecholate','a-hydroxycarboxylate','carboxylate','phenolate','citrate','other']]

# Transponemos el df
df_db_grupos_transpuesto = df_db_grupos.transpose()
df_db_grupos_transpuesto.head()

0    1    2    3    4    5    6    7    8    9    ...  \
hydroxamate             0    0    0    0    0    3    3    3    0    0  ...   
catecholate             1    1    1    1    0    0    0    0    1    0  ...   
a-hydroxycarboxylate    0    0    0    0    0    0    0    0    0    0  ...   
carboxylate             1    1    1    1    1    0    0    0    1    1  ...   
phenolate               0    0    0    0    1    0    0    0    0    1  ...   

                      211  212  213  214  215  216  217  218  219  220  
hydroxamate             3    3    0    0    3    2    0    0    0    0  
catecholate             0    0    0    0    0    0    1    1    0    1  
a-hydroxycarboxylate    0    0    0    0    0    0    0    0    0    0  
carboxylate             0    0    0    1    0    0    2    2    2    2  
phenolate               0    0    0    2    0    0    0    0    0    0  

[5 rows x 221 columns]

In [ ]:
# Seleccionamos los compuestos que pertenecen a un solo grupo del df_db y lo clasificamos
columnas = []
for i in range(220):
  if df_db_grupos_transpuesto[i].eq(0).sum() == 6:
    columnas.append(i)

df_db_grupos_un_valor = df_db_grupos_transpuesto[columnas].transpose().idxmax(axis=1)
df_db_grupos_un_valor

5      hydroxamate
6      hydroxamate
7      hydroxamate
11     catecholate
13     hydroxamate
          ...     
209    hydroxamate
211    hydroxamate
212    hydroxamate
215    hydroxamate
216    hydroxamate
Length: 112, dtype: object

In [ ]:
# Igualmente seleccionamos los compuestos en el df_db1 y eliminamos la columna ID
df_db1_transpuesto = df_db1.transpose()
df_db1_grupos_un_valor = df_db1_transpuesto[columnas].transpose().drop(columns=['ID'])
df_db1_grupos_un_valor.head()

MaxEStateIndex MinEStateIndex MaxAbsEStateIndex MinAbsEStateIndex  \
5       14.637926      -6.027201         14.637926          1.541309   
6       14.204271      -5.464974         14.204271          1.515272   
7       14.235765      -5.588196         14.235765          1.524537   
11      13.582822      -4.208434         13.582822          0.133729   
13      13.840994      -5.437481         13.840994          1.185498   

         qed    MolWt HeavyAtomMolWt  ExactMolWt NumValenceElectrons  \
5   0.030202  782.889        724.425  782.406186                 310   
6    0.02944  726.825        672.393  726.379971                 288   
7   0.028291  740.852        684.404  740.395621                 294   
11  0.577591  346.343        328.199   346.12772                 132   
13  0.052742  636.791        584.375  636.384663                 254   

   MaxPartialCharge  ... fr_NH0 fr_NH1 fr_NH2 fr_N_O fr_amide fr_benzene  \
5          0.328347  ...      4      2      0      6        6          0   
6          0.322416  ...      3      2      1      6        5          0   
7          0.322629  ...      3      3      0      6        5          0   
11         0.293318  ...      2      2      0      2        2          1   
13         0.249829  ...      3      2      1      6        5          1   

   fr_ester fr_para_hydroxylation fr_phenol fr_phenol_noOrthoHbond  
5         1                     0         0                      0  
6         1                     0         0                      0  
7         1                     0         0                      0  
11        0                     1         2                      2  
13        0                     0         0                      0  

[5 rows x 118 columns]

In [ ]:
# Juntamos los dos df
df_db_db1 = pd.concat([df_db_grupos_un_valor, df_db1_grupos_un_valor], axis=1).rename(columns={0:'target'})
df_db_db1.head()

target MaxEStateIndex MinEStateIndex MaxAbsEStateIndex  \
5   hydroxamate      14.637926      -6.027201         14.637926   
6   hydroxamate      14.204271      -5.464974         14.204271   
7   hydroxamate      14.235765      -5.588196         14.235765   
11  catecholate      13.582822      -4.208434         13.582822   
13  hydroxamate      13.840994      -5.437481         13.840994   

   MinAbsEStateIndex       qed    MolWt HeavyAtomMolWt  ExactMolWt  \
5           1.541309  0.030202  782.889        724.425  782.406186   
6           1.515272   0.02944  726.825        672.393  726.379971   
7           1.524537  0.028291  740.852        684.404  740.395621   
11          0.133729  0.577591  346.343        328.199   346.12772   
13          1.185498  0.052742  636.791        584.375  636.384663   

   NumValenceElectrons  ... fr_NH0 fr_NH1 fr_NH2 fr_N_O fr_amide fr_benzene  \
5                  310  ...      4      2      0      6        6          0   
6                  288  ...      3      2      1      6        5          0   
7                  294  ...      3      3      0      6        5          0   
11                 132  ...      2      2      0      2        2          1   
13                 254  ...      3      2      1      6        5          1   

   fr_ester fr_para_hydroxylation fr_phenol fr_phenol_noOrthoHbond  
5         1                     0         0                      0  
6         1                     0         0                      0  
7         1                     0         0                      0  
11        0                     1         2                      2  
13        0                     0         0                      0  

[5 rows x 119 columns]

### División entrenamiento-prueba

In [ ]:
# Definimos un dataframe con solo características
df_feat = df_db_db1.drop('target',axis=1)
df_feat.head()

MaxEStateIndex MinEStateIndex MaxAbsEStateIndex MinAbsEStateIndex  \
5       14.637926      -6.027201         14.637926          1.541309   
6       14.204271      -5.464974         14.204271          1.515272   
7       14.235765      -5.588196         14.235765          1.524537   
11      13.582822      -4.208434         13.582822          0.133729   
13      13.840994      -5.437481         13.840994          1.185498   

         qed    MolWt HeavyAtomMolWt  ExactMolWt NumValenceElectrons  \
5   0.030202  782.889        724.425  782.406186                 310   
6    0.02944  726.825        672.393  726.379971                 288   
7   0.028291  740.852        684.404  740.395621                 294   
11  0.577591  346.343        328.199   346.12772                 132   
13  0.052742  636.791        584.375  636.384663                 254   

   MaxPartialCharge  ... fr_NH0 fr_NH1 fr_NH2 fr_N_O fr_amide fr_benzene  \
5          0.328347  ...      4      2      0      6        6          0   
6          0.322416  ...      3      2      1      6        5          0   
7          0.322629  ...      3      3      0      6        5          0   
11         0.293318  ...      2      2      0      2        2          1   
13         0.249829  ...      3      2      1      6        5          1   

   fr_ester fr_para_hydroxylation fr_phenol fr_phenol_noOrthoHbond  
5         1                     0         0                      0  
6         1                     0         0                      0  
7         1                     0         0                      0  
11        0                     1         2                      2  
13        0                     0         0                      0  

[5 rows x 118 columns]

In [ ]:
# Definimos un dataframe con solo la variable objetivo
df_target = df_db_db1['target']
df_target.head()

5     hydroxamate
6     hydroxamate
7     hydroxamate
11    catecholate
13    hydroxamate
Name: target, dtype: object

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(df_feat, df_target, test_size=0.30, random_state=101)

In [ ]:
y_train.head()

207    catecholate
124    hydroxamate
7      hydroxamate
205    catecholate
203    hydroxamate
Name: target, dtype: object

### Entrenamos el clasificador de Soporte Vectorial

In [ ]:
from sklearn.svm import SVC

In [ ]:
model = SVC()

In [ ]:
model.fit(X_train,y_train)

SVC()

## 3. Resultados

### Predicciones y Evaluaciones

In [ ]:
predictions = model.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix

In [ ]:
print(confusion_matrix(y_test,predictions))

[[ 0  7  0]
 [ 0 26  0]
 [ 0  1  0]]


In [ ]:
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

 catecholate       0.00      0.00      0.00         7
 hydroxamate       0.76      1.00      0.87        26
   phenolate       0.00      0.00      0.00         1

    accuracy                           0.76        34
   macro avg       0.25      0.33      0.29        34
weighted avg       0.58      0.76      0.66        34



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## 4. Discusion y conlusiones
- Rendimiento desequilibrado: El modelo tiene un rendimiento deficiente en la identificación de las categorías "catecolato" y "fenolato". La precisión, el recall y el F1-score son 0.00 para ambas, lo que indica que el modelo no predice correctamente ninguna instancia de estas clases.
- Buen rendimiento en "hidroxamato": Por otro lado, el modelo funciona bien para la clase "hidroxamato". La precisión, el recall y el F1-score son todos altos (0.76, 1.00 y 0.87 respectivamente), lo que indica que el modelo identifica correctamente la mayoría de las instancias de esta clase y no hay muchas predicciones falsas positivas.
- Baja precisión general: A pesar del buen rendimiento en "hidroxamato", la precisión general del modelo es solo del 76%. Esto se debe principalmente al desequilibrio en la clasificación de las otras categorías.

## 5. Referencias

- Mijaljica D, Prescott M, Devenish RJ. The intriguing life of autophagosomes. Int J Mol Sci. 2012;13(3):3618-3635. doi: 10.3390/ijms13033618. Epub 2012 Mar 19. PMID: 22489171; PMCID: PMC3317731.